File: Week6/sentimental-credit/credit.py
# TODO
import re
import math
def double(n):
    return 2 * int(n)
while True:
    try:
        num = input("Number: ")
        if int(num):
            str(num)
            break
    except:
        print("invalid number")
sums = list(map(double, num[len(num)-2::-2]))

start = int(num[0:2])

for i in range(len(sums)):
    if sums[i] > 9:
        sums[i] = sum(int(digit) for digit in str(sums[i]))

finalSum = sum(sums)
finalSum += sum(map(int, num[len(num)-1::-2]))
#print(finalSum)

if finalSum % 10 == 0:
    #do thing
    MC = [51, 52, 53, 54, 55]
    if len(num) == 15 and (start == 37 or start == 34):
        print("AMEX")
    elif (len(num) == 16) and (start in MC):
        print("MASTERCARD")
    elif (len(num) == 13 or len(num) == 16) and (int(str(start)[0]) == 4):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

#print(start)
#print(num[len(num)-2::-2])
#print(sums)
#Luhn's algorithm

File: Week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long num, temp, startDig;
    int counter = 0;
    int sum = 0;
    num = get_long("Number: ");
    temp = num;
    startDig = num;
    temp /= 10;
//finds every other digit starting from the second to last and multiplies it by two, then adds it to sum
    while(temp){
        int tmp = 2 * (temp % 10);
        while(tmp){
            sum += tmp % 10;
            tmp /=10;
        }
        temp /= 100;
    }
//adds the remaining digit
    while(num){
        sum += num % 10;
        num /= 100;
    }
    if(sum % 10 == 0 ){
        //finds the starting digits
        while(startDig > 99){
            startDig/=10;
            //finds how many digits there are
            counter++;
        }
        //adds two to digit count cuz startDig stops at 2 digits
        counter += 2;
        if(counter == 15 || counter == 16 || counter == 13){
            if(startDig > 39 && startDig < 50){
                printf("VISA\n");
            }
            else if(startDig == 51 || startDig == 52 || startDig == 53 || startDig == 54 || startDig == 55){
                printf("MASTERCARD\n");
            }
            else if(startDig == 34 || startDig == 37){
                printf("AMEX\n");
            }
            else{
                printf("INVALID\n");
            }
        }
        else{
            printf("INVALID\n");
        }

    }
    else{
        printf("INVALID\n");
    }

}